flink-staging/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/RowComparator.scala

```diff
@@ -0,0 +1,468 @@
+        normalizableKeyPrefixLen += len + 1 // add one for a null byte
+        if (normalizableKeyPrefixLen < 0) {
+          // overflow, which means we are out of budget for normalized key space anyways
+          normalizableKeyPrefixLen = Integer.MAX_VALUE
+          return
+        }
+      }
+      else {
+        return
+      }
+      i += 1
+    }
+  }
+
+  private def instantiateDeserializationUtils(): Unit = {
+    deserializedKeyFields1 = new Array[Any](serializers.length)
+    deserializedKeyFields2 = new Array[Any](serializers.length)
+
+    var i = 0
+    while (i < serializers.length) {
+      deserializedKeyFields1(i) = serializers(i).createInstance()
+      deserializedKeyFields2(i) = serializers(i).createInstance()
+      i += 1
+    }
+  }
+
+  // --------------------------------------------------------------------------------------------
+  //  Comparator Methods
+  // --------------------------------------------------------------------------------------------
+
+  override def getFlatComparator(flatComparators: util.List[TypeComparator[_]]): Unit =
+    comparators.foreach {
+      case ctc: CompositeTypeComparator[_] => ctc.getFlatComparator(flatComparators)
+      case c@_ => flatComparators.add(c)
+    }
+
+  override def compareToReference(referencedComparator: TypeComparator[Row]): Int = {
+    val other: RowComparator = referencedComparator.asInstanceOf[RowComparator]
+    var i = 0
+    try {
+      while (i < keyPositions.length) {
+        val comparator = comparators(i).asInstanceOf[TypeComparator[Any]]
+        val otherComparator = other.comparators(i).asInstanceOf[TypeComparator[Any]]
+
+        val nullCheck = comparator.equalToReference(nullChecker)
+        val nullCheckOther = otherComparator.equalToReference(nullChecker)
+
+        var cmp = 0
+        // both values are null -> equality
+        if (nullCheck && nullCheckOther) {
+          cmp = 0
+        }
+        // one value is null -> inequality
+        // order is considered for basic types
+        else if (nullCheck || nullCheckOther) {
+          if (comparator.isInstanceOf[BasicTypeComparator[_]]) {
+            val basicComp = comparator.asInstanceOf[BasicTypeComparator[_]]
+            if (nullCheck) {
+              return if (basicComp.isAscendingComparison) 1 else -1
+            }
+            else if (nullCheckOther) {
+              return if (basicComp.isAscendingComparison) -1 else 1
+            }
+          }
+          else {
+            return if (nullCheck) 1 else -1
+          }
+        }
+        else {
+          cmp = comparator.compareToReference(otherComparator)
+        }
+
+        if (cmp != 0) {
+          return cmp
+        }
+        i = i + 1
+      }
+      0
+    }
+    catch {
+      case npex: NullPointerException =>
+        throw new NullKeyFieldException(keyPositions(i))
+      case iobex: IndexOutOfBoundsException =>
+        throw new KeyFieldOutOfBoundsException(keyPositions(i))
+    }
+  }
```

Comment:
Yeah I just thought about it myself. I guess with the zipping and so on, this is not the most efficient implementation. Therefore, I've also deleted this comment here.